//   Copyright 2011-2015 Quickstep Technologies LLC.
//   Copyright 2015 Pivotal Software, Inc.
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.

syntax = "proto2";

package quickstep.serialization;

message Type {
  enum TypeID {
    INT = 0;
    LONG = 1;
    FLOAT = 2;
    DOUBLE = 3;
    CHAR = 4;
    VAR_CHAR = 5;
    DECIMAL = 6;
    DATETIME = 7;
    DATETIME_INTERVAL = 8;
    YEAR_MONTH_INTERVAL = 9;
    NULL_TYPE = 10;
  }

  required TypeID type_id = 1;
  required bool nullable = 2;

  // The convention for extension numbering is that extensions for a particular
  // TypeID should begin from (type_id + 1) * 32.
  extensions 32 to max;
}

message CharType {
  extend Type {
    // Required when type_id == CHAR.
    optional uint64 length = 160;
  }
}

message VarCharType {
  extend Type {
    // Required when type_id == VAR_CHAR.
    optional uint64 length = 192;
  }
}
